Add group hover info to chat

applenick opened the pull request:
# Add group hover info to chat

This PR resolves #603 and resolves #615 

A brief overview, now when a user hovers over a flair (if defined) information explaining what the rank is will be shown.

To add this feature I did a few things. First created a new `Flair` class in the config to hold info related to prefix/suffix including the attributes for the hover message. Then the main change was made in `ChatDispatcher`, where now instead of sending a formatted string, a component message is created and sent to players. Allowing for the hover/click events to work as intended. 

## Config
New configurations values added:
```yaml
groups:
  # A special group for server operators.
  op:
    prefix: "&6❖"
    suffix: ""
    # Display name of the group - Displayed when hovering over prefix/suffix              
    display-name: "&6Server Operator"
    # Description of group - Displayed when hovering over prefix/suffix
    description: "&7Administrator of the server"
    # URL shown below description when hovering over prefix/suffix
    click-link: "&bhttps://pgm.dev/"

    # If you want to further customize prefix/suffix flairs, provide components here to override format
    # Only use if you know what you're doing
    prefix-component: ""
    suffix-component: “"
```

## Screenshots
Example of defining just a `display-name`
![hello-1](https://user-images.githubusercontent.com/3377659/90108665-8129e880-dcff-11ea-8caf-5d45bd1fe1cf.gif)
Example of defining a `display-name` and `description`
![hello-2](https://user-images.githubusercontent.com/3377659/90108820-bb938580-dcff-11ea-8658-888f87e528a2.gif)
Example using all hover attributes including `click-link`
![hello-3](https://user-images.githubusercontent.com/3377659/90108834-bf270c80-dcff-11ea-88c1-a714da1d122c.gif)

## Suggestions
I applied the feedback given in #603 regarding the config layout, if there are any other suggestions just let me know here and I’ll be glad to adjust this. 

As always this PR was tested and all features work as intended 👍 

Signed-off-by: applenick <applenick@users.noreply.github.com>

applenick committed:
Add group hover info to chat

Signed-off-by: applenick <applenick@users.noreply.github.com>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/namedecorations/NameDecorationProvider.java:
```diff
@@ -1,11 +1,16 @@
+
+  Component getPrefixComponent(UUID uuid);
+
+  Component getSuffixComponent(UUID uuid);
```
```suggestion
  default Component getPrefixComponent(UUID uuid) {
      return TextComponent.of(getPrefix(uuid));
  }

  default Component getSuffixComponent(UUID uuid) {
      return TextComponent.of(getSuffix(uuid));
  }
```
Keeping this interface simple to implement for 3rd parties is relevant IMO, and it's an easy enough default. (Maybe do a component parsing from legacy text instead of just a plain text with the text)

Pablete1234 commented on core/src/main/java/tc/oc/pgm/namedecorations/NameDecorationRegistry.java:
```diff
@@ -15,6 +16,17 @@
+   * @param party The party this player is currently in
+   * @return The name, decorated, in component form
+   */
+  Component getDecoratedNameComponent(Player player, Party party);
```
I'd do the same as above, provide a fallback default, even if the registry isn't really something that should be reimplemented by 3rd parties often

Electroid commented:
I'd rather not invent a new format (for chat messaging). Can we use `GsonComponentSerializer` and allow users to specify any `Component` for the flair? That way they can literally customize it in infinite ways?

`flair: {text: "...", hover: { ... }}` You get the idea.

applenick commented:
So just had a few mins to test some things out.

I’m not sure if (strictly from a user-friendly point of view) we should switch this to using raw components in the config. 

Here’s the output I got from using `GsonComponentSerializer#serialize` on the sample format I included in my screenshots...
```json
{"text":"","extra":[{"text":"§6❖"}],"clickEvent":{"action":"open_url","value":"https://oc.tc/staff"},"hoverEvent":{"action":"show_text","value":{"text":"","extra":[{"text":"§6§lAdmin"},{"text":"\n"},{"text":"This is a cool hover"},{"text":"\n"},{"translate":"chat.clickLink","with":[{"text":"https://oc.tc/staff","color":"aqua","underlined":true}],"color":"dark_aqua"}]}}}
```
I can already foresee us receiving errors from users as to why their flair is not rendering properly due to them not understanding the formatting intuitively.

The three fields I outlined (display-name, description, link), IMO, should be enough customization for a feature as simple as this. However if you do wish I continue looking into using Components instead, do you have any recommendations on how the format could be simplified so the general user can easily define the different parts of a flair?

Pablete1234 commented:
I belive we should aim for the user-friendly option, and maybe add an option for power-users who wish to define their very specific format, but that should never be the way we expect regular end users to configure this.

Keep the 3 fields, make those generate the component, and, alternatively, allow configuring a prefix-component and a suffix-component wich is this tellraw format

applenick committed:
Add flair component overrides

Signed-off-by: applenick <applenick@users.noreply.github.com>

applenick commented:
Just an update, I’ve made the changes @Pablete1234 suggested. 
To reinforce I really think we should stick with a user-friendly format for this feature, but the idea to allow additional customization via an override was a great suggestion. I’ve added that and should work as intended 👍 

If there are any other recommendations just let me know and I’d be glad to adjust it.

applenick committed:
Fix rendering of TextDecorations in usernames

Signed-off-by: applenick <applenick@users.noreply.github.com>

applenick commented:
Another update:
This PR now resolves #615 

Added some detection for text decorations and handled them. There’s probably a cleaner way to handle this, but I’m stumped. Any recommendations on how to improve the handling of ChatColor -> TextColor/Decorations, would be appreciated. 

Updated Screenshot to showcase multiple formatting code being used
![fixed-decorations](https://user-images.githubusercontent.com/3377659/90370791-ceb99480-e022-11ea-8718-49f9246cb27c.png)
Admin broadcast message & tablist matched the format shown here too
